File: src/lib/tiposkripto/tests/calculator/Calculator.test.adapter.ts
// import type { ITestAdapter } from "../../src/public/CoreTypes.js";
import type { ITestAdapter } from "../../src/CoreTypes.js";
import type { ICalculatorNode } from "./Calculator.test.types.js";

export const adapter: ITestAdapter<ICalculatorNode> = {
  prepareAll: async (input, testResource, artifactory) => {
    console.log("[adapter] beforeAll called with input:", input);
    return input;
  },
  prepareEach: async (
    subject,
    initializer,
    testResource,
    initialValues,
    artifactory,
  ) => {
    console.log("[adapter] beforeEach called with subject:", subject);
    // Trust the type contract: initializer is a function that returns the store
    const calculator = initializer(subject);
    console.log("[adapter] beforeEach created calculator:", calculator);
    return calculator;
  },
  execute: async (store, whenCB, testResource, artifactory) => {
    console.log("[adapter] andWhen called with store:", store);
    // whenCB is (calculator: Calculator) => Calculator
    const result = whenCB(store);
    console.log("[adapter] andWhen result:", result);
    return result;
  },
  verify: async (store, verificationFn, testResource, artifactory) => {
    console.log("[adapter] verify called with store:", store);
    console.log("[adapter] verificationFn:", verificationFn);
    
    if (typeof verificationFn === 'function') {
      // Call verificationFn with store to perform assertion
      await verificationFn(store);
      // Return the store (truthy value) to indicate success
      return store;
    }
    return store;
  },
  cleanupEach: async (store, key, artifactory) => {
    console.log("[adapter] afterEach called with store:", store);
    return store;
  },
  cleanupAll: async (store, artifactory) => {
    console.log(
      "afterAll called, but skipping web-only storage operations in Node.js",
    );

    artifactory.writeFileSync("fizz", "buzz");

    return store;
  },
  assert: (actual: any) => {
    console.log("[adapter] assert called with actual:", actual);
    return actual;
  },
};

File: src/lib/rubeno/examples/calculator/Calculator.test.adapter.rb
require_relative '../../lib/types'

module Rubeno
  module Examples
    class CalculatorTestAdapter
      include ITestAdapter
      
      def prepare_all(input_val, test_resource, artifactory = nil)
        puts "[adapter] before_all called with input: #{input_val}"
        input_val
      end
      
      def prepare_each(subject, initializer, test_resource, initial_values, artifactory = nil)
        puts "[adapter] beforeEach called with subject: #{subject}"
        # Call initializer with appropriate arguments
        calculator = nil
        if initializer.arity == 0
          calculator = initializer.call
        elsif initializer.arity == 1
          # Try subject first (for implementations like (input: typeof Calculator) => new input())
          calculator = initializer.call(subject)
        else
          # Default: call with no arguments
          calculator = initializer.call
        end
        puts "[adapter] beforeEach created calculator: #{calculator}"
        calculator
      end
      
      def execute(store, action_cb, test_resource, artifactory = nil)
        puts "[adapter] andWhen called with store: #{store}"
        # action_cb is (calculator: Calculator) => Calculator
        result = action_cb.call(store)
        puts "[adapter] andWhen result: #{result}"
        result
      end
      
      def verify(store, check_cb, test_resource, artifactory = nil)
        puts "[adapter] verify called with store: #{store}"
        puts "[adapter] verificationFn: #{check_cb}"
        
        if check_cb.respond_to?(:call)
          # Call verificationFn with store to perform assertion
          check_cb.call(store)
          # Return the store (truthy value) to indicate success
          return store
        end
        store
      end
      
      def cleanup_each(store, key, artifactory = nil)
        puts "[adapter] afterEach called with store: #{store}"
        store
      end
      
      def cleanup_all(store, artifactory = nil)
        puts "[adapter] afterAll called, but skipping web-only storage operations in Ruby"
        
        if artifactory && artifactory.respond_to?(:write_file_sync)
          artifactory.write_file_sync("fizz", "buzz")
        end
        
        store
      end
      
      def assert(actual)
        puts "[adapter] assert called with actual: #{actual}"
        actual
      end
    end
  end
end
